snowpark_style_guide.md

```diff
@@ -0,0 +1,413 @@
+
+```python
+# without parameter name
+df = df.join(another_df, 'id', 'inner')
+df = df.unpivot("sales", "month", ["jan", "feb"]).sort("empid")
+
+# with parameter name
+df = df.join(another_df, using_columns='id', how='inner')
+df = df.unpivot(value_column="sales", name_column="month", column_list=["jan", "feb"]).sort("empid")
+```
+
+# User Defined Functions (UDFs)
+
+[User Defined Functions (UDFs)][udf-service-doc] let you extend the system to perform customized operations and Snowpark
+lets you use the Python programming language to manipulate data and return either scalar or tabular results.
+
+UDFs can be registered in several ways and each way has its own advantages and limitations.
+
+
+## Registration
+
+**Calling the `snowflake.snowpark.functions.udf` function**
+
+The `snowflake.snowpark.functions.udf` function will register the function within the most
+recently created session. If there is only one single active session in the program, it will use that session
+for registration. However, if there are multiple active sessions in your
+program and a different session is wanted, the `session` parameter has to be explicitly set.
+In this case, you may need to keep a record of all the sessions and maintain extra logic to decide which
+session to use. It is recommended to explicitly specify the session parameter when registering.
+
+```python
+from snowflake.snowpark.functions import udf
+from snowflake.snowpark.types import IntegerType
+
+def add(x, y):
+     return x + y
+
+# without session parameter, the most recently created session will be used
+udf(
+   add,
+   return_type=IntegerType(),
+   input_types=[IntegerType(), IntegerType()]
+)
+
+# explicitly specify an active session
+udf(
+   add,
+   return_type=IntegerType(),
+   input_types=[IntegerType(), IntegerType()],
+   session=active_session
+)
+```
+
+**Using the `snowflake.snowpark.functions.udf` as decorator**
+
+`snowflake.snowpark.functions.udf` can be also used as a decorator to annotate the function to be registered.
+With the decorator, it is more clear that the function is a UDF.
+
+```python
+# udf as decorator
+from snowflake.snowpark.functions import udf
+from snowflake.snowpark.types import IntegerType
+
+@udf(
+   return_type=IntegerType(),
+   input_types=[IntegerType(), IntegerType()]
+)
+def add(x, y):
+     return x + y
+```
+
+**Calling the `snowflake.snowpark.Session.udf.register` method**
+
+The `udf` method on `Session` object returns `UDFRegistration` object which can be used for registration.
+Inherently, `UDFRegistration` is bound with the `Session` object which means the UDF will be registered
+through that session.
+
+```python
+from snowflake.snowpark.types import IntegerType
+
+def add(x, y):
+     return x + y
+
+session.udf.register(
+   add,
+   return_type=IntegerType(),
+   input_types=[IntegerType(), IntegerType()]
+)
+```
+
+## Pandas(Vectorized) UDFs
```

Comment:
Should there be a space after "Pandas"? In other words, should it be "Pandas (Vectorized) UDFs"?